### PR TITLE
added ability to create shareable link, specs

### PR DESCRIPTION
**Pull request by ShevchukTania**
**STATUS**
READY

**DESCRIPTION**

**CHECK LIST**
- [x] specs added
- [x] specs passed
- [ ] code coverage >90%
- [x] DB indexes added if needed
- [x] seed.rb file is updated if needed
- [x] PR is reviewed manually again (to make sure you have 100% ready code)
- [x] PR doesn't have conflicts with main branch
- [x] I've checked new feature as logged in and logged out user if needed
- [x] "let!" is used in specs

image

### CODE REVIEW

File: app/controllers/account/members_controller.rb [hunk 1]
```diff
@@ -26,6 +26,29 @@ def destroy
     end
   end
 
+  def greeting_new_member
+    @invitation = Invitation.find_by(token: params[:token], workspace_id: params[:workspace_id])
+
+    if @invitation.present? && @invitation.created_at >= 14.days.ago
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb]: create method "expired?" and use in 2 places here and below

File: app/controllers/account/workspaces_controller.rb [hunk 2]
```diff
@@ -39,6 +39,18 @@ def destroy
     redirect_to account_workspaces_path
   end
 
+  def create_invitation_link
+    workspace_id = params[:workspace_id]
```
Comment (igorkasyanchuk) [inline app/controllers/account/workspaces_controller.rb]: so is it possible to invite any user to any workspace? just by changing URL?

File: app/controllers/account/workspaces_controller.rb [hunk 3]
```diff
@@ -39,6 +39,18 @@ def destroy
...
+    token = Devise.friendly_token
+    @invitation = Invitation.new(invitor_id: current_user.id, workspace_id: workspace_id, token: token)
```
Comment (igorkasyanchuk) [inline app/controllers/account/workspaces_controller.rb]: more readable is to write

Invitation.new(invitor: current_user workspace: workspace, token: token)

File: app/models/invitation.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+#
+#  id           :bigint(8)        not null, primary key
+#  token        :string
+#  created_at   :datetime         not null
+#  updated_at   :datetime         not null
+#  invitor_id   :integer
+#  workspace_id :integer
+#
+
+class Invitation < ApplicationRecord
```
Comment (igorkasyanchuk) [inline app/models/invitation.rb:13]: where are associations?

Comment (igorkasyanchuk) [reply app/models/invitation.rb:13]: and factory

File: spec/factories/invitations.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+#  token        :string
+#  created_at   :datetime         not null
+#  updated_at   :datetime         not null
+#  invitor_id   :integer
+#  workspace_id :integer
+#
+
+FactoryBot.define do
+  factory :invitation do
+    invitor_id { 1 }
```
Comment (igorkasyanchuk) [inline spec/factories/invitations.rb]: this is not correct factory

File: spec/factories/invitations.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    workspace_id { 1 }
+    token Faker::Lorem.characters(20)
```
Comment (igorkasyanchuk) [inline spec/factories/invitations.rb]: this is doesn't generate uniq tokens
try to create 2 invitations in spec

File: db/migrate/20181118201736_create_invitations.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+class CreateInvitations < ActiveRecord::Migration[5.2]
+  def change
+    create_table :invitations do |t|
+      t.integer :invitor_id
+      t.integer :workspace_id
+      t.string :token
```
Comment (igorkasyanchuk) [inline db/migrate/20181118201736_create_invitations.rb:6]: add index

File: config/initializers/bitly.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+Bitly.use_api_version_3
+
+Bitly.configure do |config|
+  config.api_version = 3
+  config.access_token = "a46423bbfbf560b37eb863ead8d89df8d9afa556"
```
Comment (igorkasyanchuk) [inline config/initializers/bitly.rb:5]: in specs - do you make actual calls to bitly service? This is wrong

stub calls

tests must pass when internet is turned off on your computer

File: app/controllers/account/workspaces_controller.rb (hunk 2 above)
Comment (igorkasyanchuk) [reply app/controllers/account/workspaces_controller.rb]: maybe only workspace to which I've access?

File: app/controllers/account/members_controller.rb [hunk 9]
```diff
@@ -26,6 +26,29 @@ def destroy
...
+      @workspace = Workspace.find(params[:workspace_id])
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb]: I can pass any workspace_id? in URL and get access? this is security hole

File: db/migrate/20181118201736_create_invitations.rb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+class CreateInvitations < ActiveRecord::Migration[5.2]
+  def change
+    create_table :invitations do |t|
+      t.integer :invitor_id
```
Comment (igorkasyanchuk) [inline db/migrate/20181118201736_create_invitations.rb]: index by columns which you are using to search

File: app/controllers/account/members_controller.rb [hunk 11]
```diff
@@ -26,6 +26,29 @@ def destroy
     end
   end
 
+  def greeting_new_member
```
Comment (AndriySolonyna) [inline app/controllers/account/members_controller.rb:30]: conditions can be simplified

redirect_to root_path, notice: 'Sorry, could not identify following link' unless @invitation

if  @invitation.expired?
  ...
else
  ...
end

File: app/controllers/account/members_controller.rb [hunk 12]
```diff
@@ -27,6 +27,32 @@ def destroy
...
+    if @invitation.expired?
+      redirect_to root_path, notice: 'Sorry, this link is no longer valid'
+    else
+      @workspace = Workspace.find(params[:workspace_id])
+      @invitor = User.find(@invitation.invitor.id)
+    end
+  end
+
+  def create_thought_link
+    @workspace = Workspace.find(params[:workspace_id])
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb]: is it possible for any user to invite user to any workspace if I change URL directly?
what about security

File: app/controllers/account/members_controller.rb [hunk 13]
```diff
@@ -27,6 +27,35 @@ def destroy
...
+      redirect_to root_path, notice: 'Sorry, this link is no longer valid'
+    else
+      @workspace = @invitation.workspace
+      @invitor = @invitation.invitor
+      @token = @invitation.token
+    end
+  end
+
+  def create_thought_link
+    if @invitation = Invitation.find_by(workspace_id: params[:workspace_id], token: params[:token])
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb:48]: so I can pass any "workspace_id" again here?

File: app/controllers/account/members_controller.rb [hunk 14]
```diff
@@ -27,6 +27,35 @@ def destroy
...
+      @workspace = @invitation.workspace
+      @shared_workspace = @workspace.shared_workspaces.build(user_id: current_user.id)
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb]: change to (user: current_user)

File: app/controllers/account/members_controller.rb [hunk 15]
```diff
@@ -27,6 +27,35 @@ def destroy
     end
   end
 
+  def greeting_new_member
+    @invitation = Invitation.find_by(token: params[:token], workspace_id: params[:workspace_id])
+
+    if !(@invitation && @invitation.workspace)
```
Comment (igorkasyanchuk) [inline app/controllers/account/members_controller.rb]: if @invitation.nil? || @invitation.workspace.nil?

**Pull request merged**